### ISSUE TITLE
Add Windows tests for multiple configurations

### ISSUE THREAD

**Issue by peterjc123** [pytorch/pytorch#18319]
## Proposal

The question originated from #18176. Since there are so many tests in Linux, why don't we add tests for Windows? Below are the configurations that could be added.

1. 32 bit CPU build
2. separate CPU build with latest VS (https://devblogs.microsoft.com/cppblog/game-performance-and-compilation-time-improvements-in-visual-studio-2019/?utm_source=wechat_session&utm_medium=social&utm_oi=48218373095424 This blog says that in VS 2019, the linking time will be optimized and it produces better code for AVX)
3. Debug build for libtorch (Somehow important because in MSVC, release and debug libraries cannot be mixed)
4. CUDA 10.1  build with latest VS (Try out some new features)

If we sort them by importance, then I think it should be 3, 2, 4, 1.

## Possible plans

1. If we have sufficient build machines, then the answer is simple. We build them all.
2. If the build machines are not sufficient
    a. We can choose them according to importance. (3 or 2 + 3)
    b. We can choose alternative CI that provide build environment for Windows, like Azure Pipelines or Appveyor. But the build time will be long (\~2hrs), so we have these options.
        (1) build for master branch at an interval (2\~3 hrs)
        (2) set them as unblocking builds
        (3) build for libtorch only

## Difficulties

1. For debug and 32-bit CPU build, it is simple because we only need to set some flags or change some env vars.
2. For the latest VS and the latest CUDA, we may have to install/update them periodically.

cc: @yf225 @ezyang

**Comment by xsacha** [comment]
I've been trying to get Windows builds to work on v1.0.1 and there's so many bug fixes I have had to backport. It is clear to me there wasn't much testing with Windows, so this is an important step.

In fact, one of the issues required me to update to CUDA 10.1. I do wonder how it worked on the nightly builds.

**Comment by peterjc123** [comment]
@xsacha Yes, but what you mention is mainly about some other problems like missing documentation. e.g. our library default to `/MD`. If you want to use `/MT`, then you should pass `-DTORCH_STATIC=1` to CMake. And the reason you need to update to CUDA 10.1 is probably brought by CUDA + VS, which has strong version constraints with each other, which is documented in README.md. Due to limited build resources, we certainly cannot cover all the products that users may choose. If there is something like docker on Linux, I think it will much more convenient for the users to build PyTorch by themselves. But this is certainly a different question.

If you have any more issues, please open them in a new issue or connect with me with email. What we want to discuss here is about CI, not user issues.

**Comment by ezyang** [comment]
We can easily scale up builds, especially if we're not going to run CUDA. (We should think more if we're going to run CUDA, but that should be OK too.) (3) is easy to do and I can setup another build matrix entry for you as soon as we push scripts to build in debug mode. (Or you can fold it into the existing build job, same way we currently build non-CUDA and CUDA all in one go.) The rest depend on whether or not we need to install a different VS; for some of them, that seems necessary, but we'll have to update the Windows AMIs in that case. I can get you access to the configs but the packer scripts will take some fiddling to get working, probably. @pietern and @yf225 are the resident experts here.

**Comment by xsacha** [comment]
@peterjc123 Thanks for that TORCH_STATIC.

@ezyang Unfortunately, CUDA is the recommended usage of pytorch and it also seems to harbour the most bugs.
Right now I'd settle for any testing of libtorch. Even if it's just on the nightly builds. I haven't seen a working nightly libtorch on Windows.

**Comment by peterjc123** [comment]
https://devblogs.microsoft.com/cppblog/msvc-backend-updates-in-visual-studio-2019-preview-2/
Good news here. OpenMP 4 is going to be supported in VS 2019. Another back point for the build with latest MSVC.

**Comment by peterjc123** [comment]
@ezyang Any updates on this? VS 2019 is officially released now.

**Comment by ezyang** [comment]
@peterjc123 Do you wanna learn how our packer scripts work? Otherwise it's waiting on one of me or @yf225 or @pietern scrabbling out enough time to try building new Windows AMIs.

**Comment by peterjc123** [comment]
@ezyang Could you tell me where the scripts are so that I can edit it or send PRs on it?

**Comment by ezyang** [comment]
I'm working on getting you access

**Comment by cheungxiongwei** [comment]
@ezyang 老乡,也可以给我一份吗?我在 windows 系统上编译也遇到一些问题.

**Comment by ezyang** [comment]
@cheungxiongwei Get it here: http://web.mit.edu/~ezyang/Public/packer-windows.zip

**Comment by peterjc123** [comment]
Now that we have enabled Windows CI on Circle. Due to its large parallelism, we could proceed much further this time. Below are the things that we could do first.
1. Separate CPU and CUDA builds. It will make the build script more readable. https://github.com/pytorch/pytorch/blob/master/.jenkins/pytorch/win-test-helpers/build_pytorch.bat#L79
2. Build and test job for more python versions, at least python latest should be included. By doing this, we could avoid errors like this to happen. https://github.com/pytorch/pytorch/pull/30809

The following are all the available options that are of low priority.
1. Debug LibTorch builds. (Already tested in nightly jobs)
2. Build and test job for more cuda versions. (Code changes related to CUDA are usually platform-agnostic)
3. 32-bit CPU builds. (Users are limited)
4. Static builds. (Not tested yet)
5. Legacy OS test builds. (Not possible in CircleCI)
6. VS latest builds. (Usually we don't need this, unless MS made some significant changes to their compiler toolchain)
7. builds with other compilers, such as GCC, Intel CC, Clang. (Not easy to fix or maintain)

cc @mingbowan

**Comment by ezyang** [comment]
@peterjc123 sounds great to me!

**Comment by peterjc123** [comment]
I've bumped importance for VS latest builds due to the situation that I faced three different kinds of build failures using three different versions of VS in the past week.

**Comment by ezyang** [comment]
Bumping priority on Windows Debug build, see https://github.com/pytorch/pytorch/pull/33251

**Comment by orionr** [comment]
+1 on @ezyang's comment. Thank you, @peterjc123!

**Comment by guyang3532** [comment]
Most items on this issue have been fixed and I will close it and open some items with low priority in new issue.
Some related pr:
#37135
#38297
#38971
#39941
https://github.com/pytorch/builder/pull/442
https://github.com/pytorch/builder/pull/456
